src/planscape/forsys/generate_projects_for_a_single_scenario.R

```diff
@@ -5,84 +5,98 @@ suppressMessages({
-                                                    geo_wkt_field,
-                                                    output_scenario_name,
-                                                    output_scenario_tag) {
-  wp_str <- "weighted_priorities"
-  # Enables debug mode if output_scenario_name and output_scenario_tag are
-  # non-empty.
+generate_projects_for_a_single_scenario <- function(
+  forsys_input_data,
+  priorities,
+  priority_weights,
+  stand_id_field,
+  proj_id_field,
+  stand_area_field,
+  stand_cost_field,
+  geo_wkt_field,
+  output_scenario_name,
+  output_scenario_tag,
+  PreForsysClusterType = 'KMEANS') {
+
+  # Enables debug mode if providing output_scenario_name and output_scenario_tag
   # If enabled, data and graphs are output to directory,
   # output/<output_scenario_name>/<output_scenario_tag>/
-  enable_debug <- nchar(output_scenario_name) > 0 && 
-    nchar(output_scenario_tag) > 0
-
-  priority_impact_fields <- paste(priorities, "_PCP", sep="")
+  enable_debug <- (
+    nchar(output_scenario_name) > 0 &
+    nchar(output_scenario_tag) > 0)
 
   # Appends spm and pcp data columns for each priority.
   forsys_input_data <- forsys_input_data %>% 
-                forsys::calculate_spm(fields = priorities) %>%
-                forsys::calculate_pcp(fields = priorities)
+    forsys::calculate_spm(fields = priorities) %>%
+    forsys::calculate_pcp(fields = priorities)
+
   # Appends a preset_priority column, which contains total impact score values.
+  priority_impact_fields <- paste0(priorities, "_PCP")
+  wp_colname <- "weighted_priorities"
   suppressMessages(
     forsys_input_data <- forsys_input_data %>%
-                  forsys::combine_priorities(
-                    fields = priority_impact_fields, 
-                    weights = priority_weights, 
-                    new_field = wp_str)
-  )
+      forsys::combine_priorities(
+        fields = priority_impact_fields,
+        weights = priority_weights,
+        new_field = wp_colname))
+
   # Parses wkt in the geo_wkt column and adds it to a "geometry" column.
-  # Patchmax expects column name to be "geometry" - do not change the variable
-  # name.
-  # Internally, Patchmax calls st_sf on geometry column values. This results in
-  # an error unless st_as_sfc is called on the wkt first.
-  geometry <- lapply(forsys_input_data[geo_wkt_field], st_as_sfc)
-  forsys_input_data <- cbind(forsys_input_data, geometry)
+  forsys_input_data_formatted <- forsys_input_data %>%
+    rename("geometry" = geo_wkt_field) %>%
+    st_as_sf(wkt = "geometry")
 
-  shp <- st_as_sf(forsys_input_data)
+  # k-means clustering
+  if (PreForsysClusterType == 'KMEANS') {
+    # k-means clustering execution
+    source("forsys/spatial_clustering.R")
+    forsys_input_data_formatted <- cluster_cells_to_stands(
+      stand_data = forsys_input_data_formatted,
+      scenario_priorities = c(wp_colname, priorities))
```

Comment:
@strangeties this is something I wanted to discuss in more detail. I see you're using `weighted_priorities` for the `forsys` optimization instead of individual priorities. This makes implementation easier because we don't have to worry about the weights during the clustering. However, there may potentially be a big difference in the results of the clustering depending on whether we cluster first and then sum scores, or sum scores first and then cluster. Does this make sense? Perhaps it's easier to talk though on a GVC?

Reply:
yea - it makes sense.

I've added reasons for why I chose the feature vector I did, and why that includes weights: https://github.com/OurPlanscape/Planscape/pull/581#issuecomment-1451279866

I think it's a slightly arbitrary decision, and we don't have to fully agree as long as we document our methods, reasoning, and results - emphasis on testing the methods and getting empirical results. Trying to come to an agreement delays getting results.